sentinel-adapter/sentinel-api-gateway-adapter-common/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/common/rule/GatewayRuleManager.java

```diff
@@ -46,9 +39,9 @@
     /**
      * Gateway flow rule map: (resource, [rules...])
      */
-    private static final Map<String, Set<GatewayFlowRule>> GATEWAY_RULE_MAP = new ConcurrentHashMap<>();
+    private static final Map<String, Set<GatewayFlowRule>> GATEWAY_RULE_MAP = Collections.synchronizedMap(new LinkedHashMap<String, Set<GatewayFlowRule>>());
```

Comment:
`Collections.synchronizedMap` could bring bad performance, which is not recommended. 

The sorting could be a problem, but I think it's okay in the dashboard as it's only a sample here. In production we could store the rules in DB and sort by its id.

Reply:
Indeed the performance of `Collections.synchronizedMap` is weaker,  how about we consider that the add or modify rules are not frequent operations, maynot need high concurrency performance.
Though in memery store, user may more care about friendly visibility, the order in list page may be important.

Yes, if we store the rules in DB and sort by its id, there is no such problem.

I'm also wondering why JUC didn't offer `ConcurrentLinkedHashMap`.
Google's `ConcurrentLinkedHashMap` and JDK's `ConcurrentSkipListMap` seems also not suitable for such a circumstance.
Is there any better way?

Reply:
Or why use it here?
Why doesn't ConcurrentHashMap meet the purpose?

For changeable iterating you can load it into a CopyOnWriteSet before iterating and iterate the CopyOnWriteSet safely.
For modification both to modify in place or to initialize new container and replace the reference shared are safe.

Reply:
`Collections.synchronizedMap` will also lock the read operations (e.g. `get`), so it might not be a good solution here.

In fact, I don't think the sorting problem should be addressed in SDK (as for the SDK it's not needed till now). It might be more appropriate to resolve this in the dashboard (though in the dashboard without DB it's hard to achieve).

Reply:
@jasonjoo2010  Test add or update one gateway flow rule, when I click save, I found that the order of rules in list page are changed often, this may make user confused. It seems due to the none order of ConcurrentHashMap here. Hope it keeps in order when add or modify.

Reply:
> @jasonjoo2010 Test add or update one gateway flow rule, when I click save, I found that the order of rules in list page are changed often, this may make user confused. It seems due to the none order of ConcurrentHashMap here. Hope it keeps in order when add or modify.

Oh I see.
I think for me maybe I will fetch the keys and sort them before rendering. And if `Collections.synchronizedMap` could solve it it's better to drop a comment here explaining why use synchronized implementation but not for synchronized reason.

Reply:
ConcurrentHashMap and HashSet in RuleManager class have been reverted.
I also found that other rule like flow,degrade,system, in their list page, the same problem exists.
It doesn't has much effect on function, only effect the display.

There may be other approaches that we can think about to achieve right sorting.